File: Player.py
import random
from Deck import Deck
from typing import List, Union
from Card import MonsterCard, SpellCard, TrapCard

class Player:
    def __init__(self, name):
        self.deck = Deck.create_deck()
        self.name = name
        self.hand = []
        self.field: List[List[Union[MonsterCard, Union[SpellCard, TrapCard]]]] = [
            [None, None, None],
            [None, None, None]
        ]
        self.life = 4000
    
    def showLife(self):
        return self.life

    def format_card(self, card, row_index):
        from Position import Position
        if card is None:
            return "[          ]" 
        if card.getPosition() == Position.FACE_DOWN.value:
            placeholder = "M" if row_index == 0 else "C"
            return f"[{placeholder:^10}]"
        if isinstance(card, MonsterCard):
            return f"[{card.name:^10} - {card.getAttack()} ATK / {card.getDefense()} DEF]"
        return f"[{card.name:^10}]"

    def display_field(self, isOpponent=False):
        field_matrix = []
        for row_index, row in enumerate(self.field):
            field_matrix.append([
                self.format_card(cell, row_index) for cell in row
            ])
        monsters_row = " ".join(field_matrix[0])
        spells_traps_row = " ".join(field_matrix[1])
        if isOpponent:
            return f"\n{spells_traps_row}\n{monsters_row}\n"
        else:
            return f"\n{monsters_row}\n{spells_traps_row}\n"

    def draw_card(self):
        card = Deck.draw_card(self.deck)
        self.hand.append(card)
        print(f"{self.name} ha robado una carta.")

    def play_card(self, card_index):
        card = self.hand.pop(card_index)
        if isinstance(card, MonsterCard):
            for i in range(3):
                if self.field[0][i] is None:
                    self.field[0][i] = card
                    if card.getPosition() == 2:
                        print(f"{self.name} coloco {card.name} en la zona de monstruos [{card.getAttack} ATK]")
                    else:
                        print(f"{self.name} coloco en la zona de monstruos")
                    return card
            print("No hay espacios disponibles")
        else:
            for i in range(3):
                if self.field[1][i] is None:
                    self.field[1][i] = card
                    print(f"{self.name} coloco una carta en la zona de magicas/trampas")                     
                    return card
    
    def take_damage(self, damage:int):
        self.life -= damage
        print(f"{self.name} ha recibido {damage} puntos de daño.")         
    
    def remove_monster_card(self, card):
        for i in range(3):
            if self.field[0][i] == card:
                self.field[0][i] = None
                print(f"{card.name} ha sido destruido")
                return
            
    def remove_magic_card(self, card):
        for i in range(3):
            if self.field[1][i] == card:
                self.field[1][i] = None
                print(f"Carta {card.name} ha sido removida")
                return   
            
    def show_hand(self):
        for i, card in enumerate(self.hand):
            if isinstance(card, MonsterCard):
                print(f"{i}: [M] {card.name} [{card.getMonsterType()}] | [{card.getAttack()} ATK / {card.getDefense()} DEF] {'Atributo: '} {card.getMonsterAttribute()}")
            elif isinstance(card, SpellCard):
                print(f'{i}: [S] {card.name} | {card.getEffect()}')
            elif isinstance(card, TrapCard):
                print(f'{i}: [T] {card.name} | {card.getEffect()}')
                
    def getWeakestMonster(self):
        temp = None
                
        for card in self.field[0]:
            if isinstance(card, MonsterCard):
                if temp is None:
                    temp = card
                elif card.getAttack() < temp.getAttack():
                    temp = card
        return temp
            
    
class Machine(Player): 
    def __init__(self):
        super().__init__("Machine")
        
    def playRandomCard(self, enemy: Player):
        from Position import Position
        cardIndex = random.randint(0, len(self.hand) - 1)
        card = self.hand.pop(cardIndex)
        if isinstance(card, MonsterCard):
            for i in range(3):
                if self.field[0][i] is None:
                    should_set_attack = False
                    for enemyCard in enemy.field[0]:
                        if enemyCard is not None and enemyCard.getPosition() == 2 and enemyCard.getAttack() <= card.getAttack():
                            should_set_attack = True
                            break
                    if should_set_attack:
                        card.setPosition(Position.FACE_UP_ATAQUE)
                        print(f"{self.name} colocó {card.name} en la zona de monstruos en posición de ataque.") 
                    else:
                        card.setPosition(Position.FACE_DOWN)
                        print(f"{self.name} colocó una carta en la zona de monstruos en posición boca abajo.") 
                    self.field[0][i] = card
                    return card
        else:
            for i in range(3):
                if self.field[1][i] is None:
                    card.setPosition(Position.FACE_DOWN)
                    self.field[1][i] = card
                    print(f"{self.name} colocó una carta en la zona de mágicas/trampas.")
                    return card
        return None

File: Player.java
package mygame;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Player {
    protected List<Card> deck;
    protected String name;
    protected List<Card> hand;
    protected Card[][] field; // 2 rows x 3 columns
    protected int life;

    public Player(String name) {
        this.deck = Deck.createDeck(); 
        this.name = name;
        this.hand = new ArrayList<>();
        this.field = new Card[][] {
            { null, null, null },
            { null, null, null }
        };
        this.life = 4000;
    }

    public int showLife() {
        return life;
    }

    private String formatCard(Card card, int rowIndex) {
        if (card == null) {
            return "[          ]";
        }
        int pos = card.getPositionValue();
        if (pos == Position.FACE_DOWN.getValue()) {
            String placeholder = (rowIndex == 0) ? "M" : "C";
            return String.format("[%-10s]", placeholder);
        }
        if (card instanceof MonsterCard) {
            MonsterCard m = (MonsterCard) card;
            return "[" + String.format("%s - %d ATK / %d DEF", m.name, m.getAttack(), m.getDefense()) + "]";
        }
        return "[" + card.name + "]";
    }

    public String displayField(boolean isOpponent) {
        String[] rowStrings = new String[2];
        for (int rowIndex = 0; rowIndex < 2; rowIndex++) {
            StringBuilder sb = new StringBuilder();
            for (int colIndex = 0; colIndex < 3; colIndex++) {
                sb.append(formatCard(field[rowIndex][colIndex], rowIndex)).append(" ");
            }
            rowStrings[rowIndex] = sb.toString().trim();
        }
        // rowStrings[0] -> monsters, rowStrings[1] -> spells/traps
        if (isOpponent) {
            return "\n" + rowStrings[1] + "\n" + rowStrings[0] + "\n";
        } else {
            return "\n" + rowStrings[0] + "\n" + rowStrings[1] + "\n";
        }
    }

    public void drawCard() {
        Card card = Deck.drawCard(this.deck);
        this.hand.add(card);
        System.out.println(name + " ha robado una carta.");
    }

    public Card playCard(int cardIndex) {
        Card card = this.hand.remove(cardIndex);
        if (card instanceof MonsterCard) {
            for (int i = 0; i < 3; i++) {
                if (field[0][i] == null) {
                    field[0][i] = card;
                    MonsterCard m = (MonsterCard) card;
                    if (m.getPositionValue() == 2) {
                        System.out.println(name + " coloco " + m.name + " en la zona de monstruos [" + m.getAttack() + " ATK]");
                    } else {
                        System.out.println(name + " coloco en la zona de monstruos");
                    }
                    return card;
                }
            }
            System.out.println("No hay espacios disponibles");
        } else {
            for (int i = 0; i < 3; i++) {
                if (field[1][i] == null) {
                    field[1][i] = card;
                    System.out.println(name + " coloco una carta en la zona de magicas/trampas");
                    return card;
                }
            }
        }
        return null;
    }

    public void takeDamage(int damage) {
        this.life -= damage;
        System.out.println(name + " ha recibido " + damage + " puntos de daño.");
    }

    public void removeMonsterCard(Card card) {
        for (int i = 0; i < 3; i++) {
            if (field[0][i] == card) {
                field[0][i] = null;
                System.out.println(card.name + " ha sido destruido");
                return;
            }
        }
    }

    public void removeMagicCard(Card card) {
        for (int i = 0; i < 3; i++) {
            if (field[1][i] == card) {
                field[1][i] = null;
                System.out.println("Carta " + card.name + " ha sido removida");
                return;
            }
        }
    }

    public void showHand() {
        for (int i = 0; i < hand.size(); i++) {
            Card card = hand.get(i);
            if (card instanceof MonsterCard) {
                MonsterCard m = (MonsterCard) card;
                System.out.println(i + ": [M] " + m.name + " [" + m.getMonsterType() + "] | [" +
                                   m.getAttack() + " ATK / " + m.getDefense() + " DEF] Atributo: " +
                                   m.getMonsterAttribute());
            } else if (card instanceof SpellCard) {
                SpellCard s = (SpellCard) card;
                System.out.println(i + ": [S] " + s.name + " | " + s.getEffect());
            } else if (card instanceof TrapCard) {
                TrapCard t = (TrapCard) card;
                System.out.println(i + ": [T] " + t.name + " | " + t.getEffect());
            }
        }
    }

    public MonsterCard getWeakestMonster() {
        MonsterCard temp = null;
        for (int i = 0; i < 3; i++) {
            if (field[0][i] instanceof MonsterCard) {
                MonsterCard current = (MonsterCard) field[0][i];
                if (temp == null || current.getAttack() < temp.getAttack()) {
                    temp = current;
                }
            }
        }
        return temp;
    }
}

class Machine extends Player {
    public Machine() {
        super("Machine");
    }

    public Card playRandomCard(Player enemy) {
        Random rand = new Random();
        if (hand.isEmpty()) {
            return null;
        }
        int cardIndex = rand.nextInt(hand.size());
        Card card = hand.remove(cardIndex);
        
        if (card instanceof MonsterCard) {
            MonsterCard monster = (MonsterCard) card;
            for (int i = 0; i < 3; i++) {
                if (field[0][i] == null) {
                    boolean shouldSetAttack = false;
                    for (int j = 0; j < 3; j++) {
                        Card enemyCard = enemy.field[0][j];
                        if (enemyCard instanceof MonsterCard) {
                            MonsterCard enemyMonster = (MonsterCard) enemyCard;
                            if (enemyMonster.getPositionValue() == Position.FACE_UP_ATAQUE.getValue()
                                && enemyMonster.getAttack() <= monster.getAttack()) {
                                shouldSetAttack = true;
                                break;
                            }
                        }
                    }
                    if (shouldSetAttack) {
                        monster.setPosition(Position.FACE_UP_ATAQUE);
                        System.out.println(name + " colocó " + monster.name + 
                                           " en la zona de monstruos en posición de ataque.");
                    } else {
                        monster.setPosition(Position.FACE_DOWN);
                        System.out.println(name + 
                                           " colocó una carta en la zona de monstruos en posición boca abajo.");
                    }
                    field[0][i] = monster;
                    return monster;
                }
            }
        } else {
            for (int i = 0; i < 3; i++) {
                if (field[1][i] == null) {
                    card.setPosition(Position.FACE_DOWN);
                    field[1][i] = card;
                    System.out.println(name + " colocó una carta en la zona de mágicas/trampas.");
                    return card;
                }
            }
        }
        return null;
    }
}